### ISSUE TITLE
QoS for k3s pods. Eg. Shouldn't coreDns have QoS class 'Guaranteed'

### ISSUE THREAD

**Issue by Timvissers** [k3s-io/k3s#2564]
**Is your feature request related to a problem? Please describe.**
Yes.
K3s included pods are traefik, coredns, metrics-server, local-path-provisioner. Some of these are cluster critical, meaning that they are essential for a cluster to work. For this reason, it would be a good idea for them to take precedence in scheduling logic, eviction logic and preemption logic.

It would be great for these kind of cluster critical pods to be..:
1. Scheduling logic: to be scheduled before pods that are considered a lower priority for the cluster administrator.
2. Preemption logic: to be not preempted AND to be able to preempt pods considered lower priority
3. Eviction logic: to be not evicted when other pods considered lower priority can be evicted

**Currently in k3s**
Scheduling: To be ranked higher compared to lower priority pods in the scheduling queue. Scheduling order is based on PriorityClass, so this will be solved once https://github.com/rancher/k3s/pull/2558 is merged and released.
Preemption: Similar to Scheduling
**Eviction**: Eviction is based on Priority Class as well of QoS. Here is eviction ranking algorithm explained: https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#evicting-end-user-pods. This is currently not reflected for these pods in k3s

**Describe the solution you'd like**
As an example for CoreDns:
CoreDns currently has QoS 'Burstable' in k3s. This means that they will be evicted earlier than 'Guaranteed' pods according to the above eviction ranking. I think a coreDns pod should be Guaranteed, to ensure it will be in a lower position in the eviction queue. **To do this, it should resource requests equals to resource limits.**
The same holds for traefik, metrics-server and local-path-provisioner, at least to some extent.
What is your idea about this? 
What is the reason a pod like coreDns is not made 'Guaranteed'?

**Describe alternatives you've considered**
* do nothing. But this means eg coreDns can be evicted, which is not desirable. Especially in cases where coreDns is consuming more than it requested (which can be normal)
* locally patch the resources requests/limits and restart the cluster. But having this included in k3s is much better.
* k3s to provide a way to leave it up to the user. There is a recent ticket about this: https://github.com/rancher/k3s/issues/2559 but I think it's better to have this out of the box.

**Links**
* https://github.com/rancher/k3s/issues/2528 which solves the problem for preemption and scheduling ranking
* https://github.com/rancher/k3s/issues/2559 custom resources for k3s pods

**Comment by brandond** [comment]
#2558 has been merged to master

**Comment by ShylajaDevadiga** [comment]
Validated priorityClass in #2528. We do not set the coredns qosClass to Guaranteed in this PR. Closing this issue as resource request/limits is tracked separately in #2559 
priorityClassName for system critical pods are set as below.
```
kubectl get pods -n kube-system  -o custom-columns=NAME:.metadata.name,PRIORITY:.spec.priorityClassName|grep system
local-path-provisioner-7c458769fb-7rmd2   system-node-critical
metrics-server-86cbb8457f-d6bj5           system-node-critical
coredns-7dc8bdd58c-xrw28                  system-cluster-critical
traefik-6f9cbd9bd4-kxbbm                  system-cluster-critical
```
